### PR TITLE
KAFKA-5485: Streams should not suspend tasks twice

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @enothereska @bbejeck @dguy @guozhangwang

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5549/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5535/
Test PASSed (JDK 8 and Scala 2.12).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 1]
```diff
@@ -387,33 +387,8 @@ private void closeTopology() {
...
-        RuntimeException firstException = null;
-        try {
-            suspend(clean);
-        } catch (final RuntimeException e) {
-            clean = false;
-            firstException = e;
-            log.error("{} Could not close task: ", logPrefix, e);
-        }
-
+    // helper to avoid calling suspend() twice if a suspended task gets not reassigned and closed
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: nit: ... suspended task is not reassigned

**Review by dguy — COMMENTED** [review]
Thanks @mjsax any tests we can add for this?

**Comment by guozhangwang** [comment]
LGTM except @dguy 's comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 2]
```diff
@@ -450,6 +425,36 @@ public void close(boolean clean) {
...
+         *     - commit offsets
+         * - close state
+         *   - if (clean) write checkpoint
+         * - if (eos) close producer
+         * </pre>
+         * @param clean shut down cleanly (ie, incl. flush and commit) if {@code true} --
+         *              otherwise, just close open resources
+         */
+    @Override
+    public void close(boolean clean) {
```
Comment (enothereska) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:448]: We added 1 line to this right? I don't know why the diff shows such a large change...Actually nevermind, I see the rest of the code now.

**Review by enothereska — APPROVED** [review]

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5705/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5719/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by mjsax** [comment]
Updated this.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5708/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5722/
Test PASSed (JDK 7 and Scala 2.11).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 3]
```diff
@@ -360,7 +360,8 @@ public void suspend() {
      *   - commit offsets
      * </pre>
      */
-    private void suspend(final boolean clean) {
+    // visible for testing
+    void suspendImpl(final boolean clean) {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: why `suspendImpl`? That is a pretty horrible name for a method. Similarly `commitImpl`

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: To align with `commitImpl` -- I like the names :)

Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: The names are terrible. You don't suspendImpl something, you suspend it. You don't commitImpl, you commit.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Seems you have a pretty strong opinion on this :) Updated the PR.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5817/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5803/
Test PASSed (JDK 8 and Scala 2.12).

**Review by dguy — APPROVED** [review]
Thanks @mjsax LGTM

**Comment by dguy** [comment]
Merged to trunk and cherry-picked to 0.11.0

**Pull request closed without merging**